src/UsgsAstroLsSensorModel.cpp

```diff
@@ -2136,15 +2139,121 @@ void UsgsAstroLsSensorModel::computeLinearApproximation(
+
+    if (ip.samp < 0.0) ip.samp = 0.0;
+    if (ip.samp > numCols) ip.samp = numCols;
+    
+    MESSAGE_LOG("Computing initial guess with projective approximation");
+  } else {
+    ip.line = m_nLines / 2.0;
+    ip.samp = m_nSamples / 2.0;
+    MESSAGE_LOG("Computing initial guess with constant approx line/2 and sample/2");
+  }
+}
+
+//***************************************************************************
+// UsgsAstroLineScannerSensorModel::createProjectiveApproximation
+//***************************************************************************
+void UsgsAstroLsSensorModel::createProjectiveApproximation() {
+  MESSAGE_LOG("Calculating createProjectiveApproximation");
+
+  // Use 9 points (9*4 eventual matrix rows) as we need to fit 14 variables.
+  const int numPts = 9;
+  double u_factors[numPts] = {0.0, 0.0, 0.0, 0.5, 0.5, 0.5, 1.0, 1.0, 1.0};
+  double v_factors[numPts] = {0.0, 0.5, 1.0, 0.0, 0.5, 1.0, 0.0, 0.5, 1.0};
+
+  csm::EcefCoord refPt = getReferencePoint();
+
+  double desired_precision = 0.01;
+  double height = computeEllipsoidElevation(
+      refPt.x, refPt.y, refPt.z, m_majorAxis, m_minorAxis, desired_precision);
+  if (std::isnan(height)) {
+    MESSAGE_LOG(
+        "createProjectiveApproximation: computeElevation of"
+        "reference point {} {} {} with desired precision"
+        "{} returned nan height; nonprojective",
+        refPt.x, refPt.y, refPt.z, desired_precision)
+    m_useApproxInitTrans = false;
+    return;
+  }
+  MESSAGE_LOG(
+      "createProjectiveApproximation: computeElevation of"
+      "reference point {} {} {} with desired precision"
+      "{} returned {} height",
+      refPt.x, refPt.y, refPt.z, desired_precision, height)
+
+  double numImageRows = m_nLines;
+  double numImageCols = m_nSamples;
+
+  std::vector<csm::ImageCoord> ip(2 * numPts);
+  std::vector<csm::EcefCoord>  gp(2 * numPts);
+
+  // Sample at two heights above the ellipsoid in order to get a
+  // reliable estimate of the relationship between image points and
+  // ground points.
+  
+  for (int i = 0; i < numPts; i++) {
+    ip[i].line = u_factors[i] * numImageRows;
+    ip[i].samp = v_factors[i] * numImageCols;
+    gp[i]      = imageToGround(ip[i], height);
+  }
+
+  double delta_z = 100.0;
+  height += delta_z;
+  for (int i = 0; i < numPts; i++) {
+    ip[i + numPts].line = u_factors[i] * numImageRows;
+    ip[i + numPts].samp = v_factors[i] * numImageCols;
+    gp[i + numPts]      = imageToGround(ip[i + numPts], height);
+  }
+
+  usgscsm::computeBestFitProjectiveTransform(ip, gp, m_projTransCoeffs);
+  m_useApproxInitTrans = true;
+  
+  MESSAGE_LOG("Completed createProjectiveApproximation");
+}
+
+//***************************************************************************
+// UsgsAstroLineScannerSensorModel::createLinearApproximation
+//***************************************************************************
+void UsgsAstroLsSensorModel::createLinearApproximation() {
```

Comment:
This function looks like solid code which performs very well. It could be of some use perhaps, one day, or if the proposed projective transform has issue for some quirky sensor (I don't anticipate that, but who knows).  I could factor this out to Utilities.cpp, together with the formula for the determinant, those are too low-level to keep in this sensor's implementation.

Reply:
In retrospect, I can wipe this old code. It looks like too much work to refactor it. If we need a best-fit linear transform going forward, the code for the projective transform can be tweaked a little to exclude the denominators, and then that will give a linear transform.